localads/campaign_serialization.cpp

```diff
@@ -0,0 +1,84 @@
+#include "localads/campaign_serialization.hpp"
+
+#include "coding/varint.hpp"
+#include "coding/byte_stream.hpp"
+
+#include <vector>
+#include <type_traits>
+#include <cstdint>
+
+
+namespace
+{
+enum class Version
+{
+  unknown = -1,
+  v1 = 0,  // March 2017 (Store featrues ids as varints, use two bytes for
+           // both days before expiration and icon id.)
+  latest = v1
+};
+
+template<typename ByteStream, typename T>
+void Write(ByteStream & s, T t)
+{
+  s.Write(&t, sizeof(T));
+}
+
+template<typename T, typename ByteStream>
+T Read(ByteStream & s)
+{
+  T t;
+  s.Read(static_cast<void*>(&t), sizeof(t));
+  return t;
+}
+
+template<typename Integral,
+         typename ByteStream,
+         typename std::enable_if<std::is_integral<Integral>::value, void*>::type = nullptr>
+std::vector<Integral> ReadData(ByteStream & s, size_t chunksNumber)
+{
+  std::vector<Integral> result;
+  while (chunksNumber--)
+    result.push_back(ReadVarUint<uint32_t>(s));
+  return result;
+}
+}  // namespace
+
+namespace localads
+{
+std::vector<uint8_t> Serialize(std::vector<Campaign> const & campaigns)
+{
+  std::vector<uint8_t> buff;
+  PushBackByteSink<decltype(buff)> dst(buff);
+  Write(dst, Version::latest);
+  Write(dst, campaigns.size());
+  for (auto const & c : campaigns)
+     WriteVarUint(dst, c.m_featureId);
+  for (auto const & c : campaigns)
+    WriteVarUint(dst, c.m_iconId);
+
+  for (auto const & c : campaigns)
+    Write(dst, c.m_daysBeforeExpired);
```

Comment:
Not right no. It's a number between 1 and 30 and fits in one byte. The best way to compress it would be using some 5 free bits of other field.